// @luma.gl/shadertools/dist/lib/shader-assembly/resolve-modules.js
import { ShaderModuleInstance } from "../shader-module/shader-module-instance.js";
export function resolveModules(modules) {
  const instances = ShaderModuleInstance.instantiateModules(modules);
  return getShaderDependencies(instances);
}
function getShaderDependencies(modules) {
  const moduleMap = {};
  const moduleDepth = {};
  getDependencyGraph({
    modules,
    level: 0,
    moduleMap,
    moduleDepth
  });
  return Object.keys(moduleDepth).sort((a, b) => moduleDepth[b] - moduleDepth[a]).map(name => moduleMap[name]);
}
export function getDependencyGraph(options) {
  const {
    modules,
    level,
    moduleMap,
    moduleDepth
  } = options;
  if (level >= 5) {
    throw new Error('Possible loop in shader dependency graph');
  }
  for (const module of modules) {
    moduleMap[module.name] = module;
    if (moduleDepth[module.name] === undefined || moduleDepth[module.name] < level) {
      moduleDepth[module.name] = level;
    }
  }
  for (const module of modules) {
    if (module.dependencies) {
      getDependencyGraph({
        modules: module.dependencies,
        level: level + 1,
        moduleMap,
        moduleDepth
      });
    }
  }
}
export const TEST_EXPORTS = {
  getShaderDependencies,
  getDependencyGraph
};
