// @luma.gl/webgl/dist/context/polyfill/polyfill-context.js
import { assert } from '@luma.gl/core';
import { polyfillVertexArrayObject } from "./polyfill-vertex-array-object.js";
import { WEBGL2_CONTEXT_POLYFILLS, WEBGL2_CONTEXT_OVERRIDES } from "./polyfill-table.js";
import { getContextData } from "./context-data.js";
export function polyfillContext(gl) {
  const contextState = getContextData(gl);
  if (!contextState._polyfilled) {
    polyfillVertexArrayObject(gl);
    initializeExtensions(gl);
    installPolyfills(gl, WEBGL2_CONTEXT_POLYFILLS);
    installOverrides(gl, {
      target: contextState,
      target2: gl
    });
    contextState._polyfilled = true;
  }
  return gl;
}
function initializeExtensions(gl) {
  const contextState = getContextData(gl);
  const EXTENSIONS = gl.getSupportedExtensions() || [];
  const IGNORE_EXTENSIONS = ['WEBGL_polygon_mode'];
  for (const extensionName of EXTENSIONS) {
    if (!IGNORE_EXTENSIONS.includes(extensionName)) {
      const extension = gl.getExtension(extensionName);
      contextState._extensions[extensionName] = extension;
    }
  }
}
function installPolyfills(gl, polyfills) {
  const contextState = getContextData(gl);
  for (const extension of Object.getOwnPropertyNames(polyfills)) {
    if (extension !== 'overrides') {
      polyfillExtension(gl, {
        extension,
        target: contextState,
        target2: gl
      });
    }
  }
}
function polyfillExtension(gl, _ref) {
  let {
    extension,
    target,
    target2
  } = _ref;
  const defaults = WEBGL2_CONTEXT_POLYFILLS[extension];
  assert(defaults);
  const {
    meta = {}
  } = defaults;
  const {
    suffix = ''
  } = meta;
  const ext = gl.getExtension(extension);
  for (const key of Object.keys(defaults)) {
    const extKey = `${key}${suffix}`;
    let polyfill = null;
    if (key === 'meta') {} else if (typeof gl[key] === 'function') {} else if (ext && typeof ext[extKey] === 'function') {
      polyfill = function () {
        return ext[extKey](...arguments);
      };
    } else if (typeof defaults[key] === 'function') {
      polyfill = defaults[key].bind(target);
    }
    if (polyfill) {
      target[key] = polyfill;
      target2[key] = polyfill;
    }
  }
}
function installOverrides(gl, _ref2) {
  let {
    target,
    target2
  } = _ref2;
  Object.keys(WEBGL2_CONTEXT_OVERRIDES).forEach(key => {
    if (typeof WEBGL2_CONTEXT_OVERRIDES[key] === 'function') {
      const originalFunc = gl[key] ? gl[key].bind(gl) : () => {};
      const polyfill = WEBGL2_CONTEXT_OVERRIDES[key].bind(null, gl, originalFunc);
      target[key] = polyfill;
      target2[key] = polyfill;
    }
  });
}
