// @luma.gl/core/dist/adapter/type-utils/decode-texture-format.js
import { decodeVertexType } from "./decode-data-type.js";
const REGEX = /^(rg?b?a?)([0-9]*)([a-z]*)(-srgb)?(-webgl|-unsized)?$/;
export function decodeTextureFormat(format) {
  const matches = REGEX.exec(format);
  if (matches) {
    const [, format, length, type, srgb, suffix] = matches;
    if (format) {
      const dataType = `${type}${length}`;
      const decodedType = decodeVertexType(dataType);
      return {
        format: format,
        components: format.length,
        srgb: srgb === '-srgb',
        unsized: suffix === '-unsized',
        webgl: suffix === '-webgl',
        ...decodedType
      };
    }
  }
  return decodeNonStandardFormat(format);
}
const EXCEPTIONS = {
  'rgba4unorm-webgl': {
    format: 'rgba',
    bpp: 2
  },
  'rgb565unorm-webgl': {
    format: 'rgb',
    bpp: 2
  },
  'rgb5a1unorm-webgl': {
    format: 'rgba',
    bbp: 2
  },
  'rgb9e5ufloat': {
    format: 'rgb',
    bbp: 4
  },
  'rg11b10ufloat': {
    format: 'rgb',
    bbp: 4
  },
  'rgb10a2unorm': {
    format: 'rgba',
    bbp: 4
  },
  'rgb10a2unorm-webgl': {
    format: 'rgba',
    bbp: 4
  },
  'stencil8': {
    components: 1,
    bpp: 1,
    a: 'stencil'
  },
  'depth16unorm': {
    components: 1,
    bpp: 2,
    a: 'depth'
  },
  'depth24plus': {
    components: 1,
    bpp: 3,
    a: 'depth'
  },
  'depth32float': {
    components: 1,
    bpp: 4,
    a: 'depth'
  },
  'depth24plus-stencil8': {
    components: 2,
    bpp: 4,
    a: 'depth-stencil'
  },
  'depth24unorm-stencil8': {
    components: 2,
    bpp: 4,
    a: 'depth-stencil'
  },
  'depth32float-stencil8': {
    components: 2,
    bpp: 4,
    a: 'depth-stencil'
  }
};
function decodeNonStandardFormat(format) {
  var _data$format;
  const data = EXCEPTIONS[format];
  if (!data) {
    throw new Error(`Unknown format ${format}`);
  }
  return {
    format: data.format || '',
    components: data.components || ((_data$format = data.format) === null || _data$format === void 0 ? void 0 : _data$format.length) || 1,
    byteLength: data.bpp || 1,
    srgb: false,
    unsized: false
  };
}
