// @luma.gl/core/dist/adapter/resources/texture.js
let _Symbol$toStringTag;
import { Resource } from "./resource.js";
_Symbol$toStringTag = Symbol.toStringTag;
export class Texture extends Resource {
  get [_Symbol$toStringTag]() {
    return 'Texture';
  }
  constructor(device, props) {
    let defaultProps = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : Texture.defaultProps;
    super(device, props, defaultProps);
    this.dimension = void 0;
    this.format = void 0;
    this.width = void 0;
    this.height = void 0;
    this.depth = void 0;
    this.sampler = void 0;
    this.dimension = this.props.dimension;
    this.format = this.props.format;
    this.width = this.props.width;
    this.height = this.props.height;
    this.depth = this.props.depth;
  }
}
Texture.defaultProps = {
  ...Resource.defaultProps,
  data: null,
  dimension: '2d',
  format: 'rgba8unorm',
  width: undefined,
  height: undefined,
  depth: 1,
  mipmaps: true,
  sampler: {},
  compressed: false,
  usage: 0,
  mipLevels: undefined,
  samples: undefined,
  type: undefined
};
Texture.COPY_SRC = 0x01;
Texture.COPY_DST = 0x02;
Texture.TEXTURE_BINDING = 0x04;
Texture.STORAGE_BINDING = 0x08;
Texture.RENDER_ATTACHMENT = 0x10;
